Copy published crossgen2 in artifacts/tests

am11 opened the pull request:
Fixes #80110

am11 commented:
cc @jkoritzinsky, @trylek 

The target `CopyDependencyToCoreRoot` in `src/tests/Directory.Build.targets` seems to be (an unused?) duplicate of the one with same name in `src/tests/Common/Directory.Build.targets`, so I have updated both. If someone can confirm it is safe to delete it from `src/tests/Directory.Build.targets` I can push a change here. Otherwise, it can be cleaned up in a later PR.

am11 commented on src/coreclr/tools/r2rtest/Crossgen2Runner.cs:
```diff
@@ -64,7 +64,7 @@ public Crossgen2Runner(BuildOptions options, Crossgen2RunnerOptions crossgen2Run
         {
             ProcessParameters processParameters = base.CompilationProcess(outputFileName, inputAssemblyFileNames);
-            processParameters.Arguments = $"{Crossgen2Path} {processParameters.Arguments}";
+            processParameters.ProcessPath = $"{Crossgen2Path}";
```
```suggestion
            processParameters.ProcessPath = Crossgen2Path;
```

trylek commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -122,7 +137,7 @@
             $(Crossgen2Dir)/jitinterface_*;
             $(Crossgen2Dir)/Microsoft.DiaSymReader.Native.*.dll"
-          TargetDir="crossgen2/" />
+          TargetDir="../crossgen2/" />
```
Can you please explain the reason for this change? According to what I see at line 305, it seems like the double period would crawl out of <code>CORE_ROOT</code>.

am11 replied:
This is producing the crossgen2 directory out of `CORE_ROOT`, then in `src/tests/build.cmd/sh` we `move` it back inside the `CORE_ROOT`. There was a problem with MSBuild mixing the files to copy selected by glob without this workaround (and I actually don't remember whether this workaround was enough for all publishing modes, will try to re-run the tests when I get some free cycles).

trylek replied:
Hmm, that sounds weird, I would rather suspect that what's actually happening is that for some reason the population of CORE_ROOT tries to copy both versions of crossgen2 and that naturally fails due to clashing duplicates. If you have trouble double-checking that in binlogs, I can take a look locally and try to figure out what's going on, we shouldn't need to copy the files around twice as you describe.

am11 replied:
It is done on line 41 via `PublishDir=$(CORE_ROOT)\..\crossgen2`. This part is in sync with that path. IIRC, the issue was with the list generated by MSBuild via glob misses some files. I was comparing the text logs generated via `-v:diag`. Will run again to refresh the memory.

trylek commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -18,6 +18,8 @@
     <!-- Non-desktop OS's use a custom dotnet host, instead of corerun -->
     <IsDesktopOS Condition="'$(TargetsBrowser)' != 'true' and '$(TargetsAndroid)' != 'true' and '$(TargetstvOS)' != 'true' and '$(TargetsiOS)' != 'true' and '$(TargetsMacCatalyst)' != 'true'">true</IsDesktopOS>
+    <Crossgen2Supported Condition="'$(RuntimeFlavor)' == 'coreclr' and '$(TestBuildMode)' != 'nativeaot'">true</Crossgen2Supported>
+    <UsePublishedCrossgen2 Condition="'$(Crossgen2Supported)' == 'true' and '$(Configuration)' != 'checked'">true</UsePublishedCrossgen2>
```
To avoid duplicating these properties between <code>src/tests/Common/D.B.targets</code> and <code>src/tests/D.B.targets</code> you could put them in <code>src/tests/D.B.props</code>, that gets included by both.

trylek commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -18,6 +18,8 @@
     <!-- Non-desktop OS's use a custom dotnet host, instead of corerun -->
     <IsDesktopOS Condition="'$(TargetsBrowser)' != 'true' and '$(TargetsAndroid)' != 'true' and '$(TargetstvOS)' != 'true' and '$(TargetsiOS)' != 'true' and '$(TargetsMacCatalyst)' != 'true'">true</IsDesktopOS>
+    <Crossgen2Supported Condition="'$(RuntimeFlavor)' == 'coreclr' and '$(TestBuildMode)' != 'nativeaot'">true</Crossgen2Supported>
+    <UsePublishedCrossgen2 Condition="'$(Crossgen2Supported)' == 'true' and '$(Configuration)' != 'checked'">true</UsePublishedCrossgen2>
```
I'm somewhat worried that by forking Crossgen2 behavior for checked vs. release mode we're making life harder for developers. According to the tracking issue it's bad behavior that we're testing our internal raw Crossgen2 build, not the final NativeAOT-compiled thing we ship to customers and I don't see any apparent reason why we should maintain it in checked build mode. (And what about debug mode?) Also, as you can see, the <code>coreclr-runtime outerloop</code> pipeline we're using as one of the main vehicles for Crossgen2 testing uses <code>Checked</code> build mode for all its R2R jobs so your change wouldn't actually translate into any "fidelity improvements" in it. We'd either need to be smarter about running some R2R jobs using checked and others using release mode but for now I'd be inclined to just remove the configuration check. We might want to only set <code>UsePublishedCrossgen2</code> when not explicitly set previously, that would let developers explicitly override it for the purpose of their local innerloop; we could then also create a new occasionally running pipeline (e.g. once a week) that would force-set the property to false so that we could keep some amount of regression testing for the "raw" Crossgen2.

trylek replied:
/cc @jkoritzinsky and @hoyosjs for additional thoughts on this

am11 replied:
I switched to other things, but this is still work-in-progress as some tests were failing. 🥲

The only reason why I added this condition was because NativeAOT'd published crossgen2 was failing some asserts. I wanted to (temporarily) get by that error to find out if there are other blockers. Unfortunately, there still are issues even in release and debug builds. Things get interesting when we run the published crossgen2 executable (AOT, singlefile or even a simple self-contained app); `crossgen2.exe <args>` fails where previous `dotnet crossgen2.dll <args>` was working.

trylek replied:
If you have a hard time figuring out the problem with NativeAOT Crossgen2 we can probably loop in @MichalStrehovsky for help but I'd be surprised if it was completely non-functional as that's literally what we ship :-).

trylek replied:
Can you please quote an example of Crossgen2 command line and the error it's reporting?

am11 replied:
Example from new CI run (now `checked` is using `PublishAot`): https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-80154-merge-7538eff8dd144b4eb2/PayloadGroup0/1/console.a9f514c2.log?helixlogtype=result
```
  Starting:    readytorun.tests.XUnitWrapper (parallel test collections = on, max threads = 6)
    readytorun/tests/mainv1/mainv1.sh [FAIL]
      Unhandled Exception: EETypeRva:0x0121FF08(System.Diagnostics.DebugProvider+DebugAssertException): GetFileSystemType returned 26
         at System.Diagnostics.DebugProvider.Fail(String, String) + 0x35
         at System.Diagnostics.Debug.Fail(String, String) + 0x24
         at Interop.Sys.TryGetFileSystemType(SafeFileHandle, Interop.Sys.UnixFileSystemTypes&) + 0xdb
         at Microsoft.Win32.SafeHandles.SafeFileHandle.CanLockTheFile(Interop.Sys.LockOperations, FileAccess) + 0x5b
         at Microsoft.Win32.SafeHandles.SafeFileHandle.Init(String, FileMode, FileAccess, FileShare, FileOptions, Int64, Int64&, UnixFileMode&) + 0x157
         at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String, FileMode, FileAccess, FileShare, FileOptions, Int64, UnixFileMode, Int64&, UnixFileMode&, Func`4) + 0x89
         at System.IO.Strategies.OSFileStreamStrategy..ctor(String, FileMode, FileAccess, FileShare, FileOptions, Int64, Nullable`1) + 0x7f
         at System.IO.FileStream..ctor(String, FileMode, FileAccess, FileShare, Int32, FileOptions, Int64) + 0x9a
         at System.IO.StreamWriter.ValidateArgsAndOpenPath(String, Boolean, Encoding, Int32) + 0x7e
         at ILCompiler.PEWriter.MapFileBuilder.SaveMap(String) + 0x5b
         at ILCompiler.DependencyAnalysis.ReadyToRunObjectWriter.EmitPortableExecutable() + 0xb77
         at ILCompiler.ReadyToRunCodegenCompilation.Compile(String) + 0x1c2
         at ILCompiler.Program.RunSingleCompilation(Dictionary`2, InstructionSetSupport, String, Dictionary`2, HashSet`1, ReadyToRunCompilerContext) + 0x1af0
         at ILCompiler.Program.Run() + 0xea4
         at ILCompiler.Crossgen2RootCommand.<>c__DisplayClass187_0.<.ctor>b__0(InvocationContext context) + 0x25b
         at System.CommandLine.Invocation.AnonymousCommandHandler.Invoke(InvocationContext) + 0x1e
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext() + 0x105
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine&) + 0x3f
         at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<BuildInvocationChain>b__0(InvocationContext invocationContext, Func`2 _) + 0x35
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseErrorReporting>b__0>d.MoveNext() + 0x9b
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine&) + 0x3f
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<UseParseErrorReporting>b__0(InvocationContext context, Func`2 next) + 0x39
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseHelp>b__0>d.MoveNext() + 0x4e
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine&) + 0x3f
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<UseHelp>b__0(InvocationContext context, Func`2 next) + 0x39
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseVersionOption>b__0>d.MoveNext() + 0x168
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine&) + 0x3f
         at System.CommandLine.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<UseVersionOption>b__0(InvocationContext context, Func`2 next) + 0x39
         at System.CommandLine.Invocation.InvocationPipeline.<Invoke>g__FullInvocationChain|3_0(InvocationContext) + 0x7c
         at System.CommandLine.Invocation.InvocationPipeline.Invoke(IConsole) + 0x81
         at ILCompiler.Program.Main(String[] args) + 0x169
         at crossgen2!<BaseAddress>+0x767205
```
this is the failing assertion: https://github.com/dotnet/runtime/blob/031a16ae6797e3ee0ebf718364a533a76b9083bc/src/libraries/Common/src/Interop/Unix/System.Native/Interop.UnixFileSystemTypes.cs#L159 and the decimal value 26 is 0x1A, which should map to apfs: https://github.com/dotnet/runtime/blob/031a16ae6797e3ee0ebf718364a533a76b9083bc/src/libraries/Common/src/Interop/Unix/System.Native/Interop.UnixFileSystemTypes.cs#L27 perhaps it was trimmed off by the linker?

MichalStrehovsky replied:
> perhaps it was trimmed off by the linker?

Yes, the enum is not public and parts of CoreLib that are not public are blocked from reflection. The concept of reflection blocking was inherited from .NET Native. It is of less importance now that our approach to reflection is different than it was in .NET Native. #72570 tracks getting rid of it. The hope is that the problem will go away eventually.

As for workarounds, we could add

```xml
<!-- Some asserts in CoreLib are not compatible with metadata blocking -->
<ItemGroup  Condition="'$(Configuration)' != 'Release'>
    <IlcArg Include="--nometadatablocking" />
</ItemGroup>
```

to crossgen2.csproj

am11 replied:
@MichalStrehovsky, thank you! Just curious, we are excluding `nometadatablocking` for Release because we are not running the tests with release configuration (and not because `nometadatablocking` is implicitly set for release), right?

trylek replied:
It may be the case that the problematic reflection constructs only affect [debug-only] assertion checks, not "actual" runtime functionality, and as we're skipping asserts in release build mode there's no reason to bloat the NativeAOT Crossgen2 app by suppressing some of the IL Linker filtering; it also makes sure that the shipping release build mode of Crossgen2 is not affected by this workaround.

trylek commented:
Overall looks good to me, thanks Adeel, in general I believe this to be heading in the right direction. We probably need to soften some rough edges though. It seems to me that quite a few R2R tests failed in your PR run, that will require investigation and fixing (theoretically these could be caused by the weird addition of ".." in populating <code>CORE_ROOT</code> I commented on in the review as I find that surprising. Please also see my comment regarding the filtering on checked build configuration. Once we're on the same page w.r.t. these aspects, I'll be more than happy to approve your change and merge it in.

am11 commented:
Remaining 12 (out of 198) failures related to R2Rdump are on linux-arm leg. corssgen2 is published with `PublishSingleFile` on linux-arm instead of PublishAot. It runs into a segmentation fault:

```sh
      BEGIN EXECUTION
      in takeLock
      15:33:18
      Response file: /root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll.rsp
      /root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/IL-CG2/crossgen2smoke.dll
      -o:/root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll
      -r:/root/helix/work/correlation/System.*.dll
      -r:/root/helix/work/correlation/Microsoft.*.dll
      -r:/root/helix/work/correlation/mscorlib.dll
      --verify-type-and-field-layout
      --method-layout:random
      --targetarch:arm
      --targetos:linux
      -O
      Running CrossGen2:  /root/helix/work/correlation/crossgen2/crossgen2 @/root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll.rsp   --composite
      Running R2RDump:  dotnet /root/helix/work/correlation/R2RDump/R2RDump.dll --header --sc --in /root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll --out /root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll.r2rdump --val
      Warning: No files matching /root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll
      Error: System.ArgumentException: Input filename must be specified (--in <file>)
         at R2RDump.Program.Run() in /_/src/coreclr/tools/r2rdump/Program.cs:line 432
      
      15:33:19
      Crossgen2 failed with exitcode: 139
```
`Running CrossGen2:` command  actually crashed with 139, and since we continue on error here, R2RDump is fails to find input file (output from crossgen2) which doesn't exist.

am11 commented:
Remaining 12 (out of 198) failures related to R2Rdump are on linux-arm leg. corssgen2 is published with `PublishSingleFile` on linux-arm instead of PublishAot. It runs into a segmentation fault:

```sh
      BEGIN EXECUTION
      in takeLock
      15:33:18
      Response file: /root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll.rsp
      /root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/IL-CG2/crossgen2smoke.dll
      -o:/root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll
      -r:/root/helix/work/correlation/System.*.dll
      -r:/root/helix/work/correlation/Microsoft.*.dll
      -r:/root/helix/work/correlation/mscorlib.dll
      --verify-type-and-field-layout
      --method-layout:random
      --targetarch:arm
      --targetos:linux
      -O
      Running CrossGen2:  /root/helix/work/correlation/crossgen2/crossgen2 @/root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll.rsp   --composite
      Running R2RDump:  dotnet /root/helix/work/correlation/R2RDump/R2RDump.dll --header --sc --in /root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll --out /root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll.r2rdump --val
      Warning: No files matching /root/helix/work/workitem/e/readytorun/crossgen2/crossgen2smoke/composite-r2r.dll
      Error: System.ArgumentException: Input filename must be specified (--in <file>)
         at R2RDump.Program.Run() in /_/src/coreclr/tools/r2rdump/Program.cs:line 432
      
      15:33:19
      Crossgen2 failed with exitcode: 139
```
`Running CrossGen2:` command  actually crashed with 139, and since we continue on error here, R2RDump fails to find input file (output from crossgen2) which doesn't exist.

trylek commented:
In our today .NET core runtime sync several people emphasized that we should coordinate this work with the diagnostics team to make sure we don't paint ourselves in a corner due to the fact that today NativeAOT bugs are in general harder to analyze and investigate. /cc-ing @tommcdon for visibility.

tommcdon commented:
@trylek please let us know how we can help with ensuring your team has the right diagnostics for Native AOT

trylek commented:
Thanks Tom for your immediate response, according to my understanding the questions raised were basically the following:

* Are we able to collect and open dumps when NativeAOT Crossgen2 fails in Helix when compiling tests?
* What is the debugging experience for NativeAOT Crossgen2 compared to the raw MSIL version?
* We should introduce lab testing similar to today <code>coreroot_determinism</code> to ensure that the raw Crossgen2 produces the same outputs as the NativeAOT Crossgen2. (This is not actually a diagnostic ask, I can do that, I just need to double-check whether it can be done before this PR has been merged in.)
* Longer term, do we track the work to collect dumps from AzDO builds to cater for the situation when NativeAOT Crossgen2 crashes when compiling the framework assemblies into <code>CORE_ROOT</code>? (Admittedly this one is not NativeAOT specific but is still important.)

Thanks

Tomas

am11 commented:
Crossgen2 in the shipping nuget package is published as AOT or SingleFile app, depending on the platform. On linux-arm (the current crash blocking this PR) is in SingleFile, as NativeAOT is not supported on linux-arm.

When one of the tests from `src/libraries` crashes, this self-explanatory template is used to print the steps about how to pull the crash dump and analyze it locally: https://github.com/dotnet/runtime/blob/main/eng/testing/debug-dump-template.md. But for the tests that run from elsewhere in the repo (`src/{tests,coreclr,mono}`) that template is unimplemented and the only choice is then to build it locally for the platform and run the tests. I am on osx-arm64 machine atm, and since cross-compilation (for linux-arm) doesn't work conveniently (without nested emulation; extremely slow), it is additional effort to run a x64 VM in cloud, cross-compile clr+libs+host+tests for arm and transfer to arm machine and repro the tests. I will eventually do that when I get some time, but this "whole lot of additional work" can be saved if the automated process is improved.

MichalStrehovsky commented:
> Are we able to collect and open dumps when NativeAOT Crossgen2 fails in Helix when compiling tests?

Like @am11 said - if the workitem execution collects the dump, we'll get the same dump we'd get if CoreCLR-hosted crossgen2 crashed instead. The problem is that currently (especially wrt the last blocker for this PR) we don't collect dumps when the crash happens on the build machine. This is irrespective of NativeAOT or not (and ironically, the issue that is happening in this PR is likely a ReadyToRun bug because it happens for the singlefile-trimmed-r2r-compiled version of crossgen2; it could also be a trimming/single file bug, but those tend to not be arm32 Linux specific).

> What is the debugging experience for NativeAOT Crossgen2 compared to the raw MSIL version?

The dump debugging experience is similar to what you're used to from .NET Native - C# with a C++-level debugging experience.

> We should introduce lab testing similar to today coreroot_determinism to ensure that the raw Crossgen2 produces the same outputs as the NativeAOT Crossgen2. (This is not actually a diagnostic ask, I can do that, I just need to double-check whether it can be done before this PR has been merged in.)

We could, but it feels a bit paranoid - would we also do this kind of testing for CoreCLR-hosted versions? We could absolutely have a comparison run that compares ReadyToRun+singlefile+trimmed crossgen2 behaves the same as a non-trimmed, non-singlefile and fully JITted one. We generally trust that R2R/SingleFile/trimming testing would catch these, even though crossgen2 is a complex codebase and it could expose unique bugs (like we see in this PR for arm32 Linux).

> Longer term, do we track the work to collect dumps from AzDO builds to cater for the situation when NativeAOT Crossgen2 crashes when compiling the framework assemblies

Not aware of such issue, but this would be super valuable elsewhere too from my past experience.

MichalStrehovsky commented:
I wonder if it would make sense to try disabling ReadyToRun for the crossgen2 configuration that is failing here and file a bug (i.e. don't block this PR on fixing that).

If we're seeing crossgen2 crashing in the CI, it is likely also crashing on customer machines since this PR is switching to testing actual customer scenarios.

If we can find a configuration that doesn't crash, switching to that configuration is a net improvement. Once we narrow it down to a readytorun/singlefile/trimming issue, the appropriate team can deal with that bug independently of this PR (and switch the configuration back once the issue is fixed).

MichalStrehovsky commented:
(I am assuming this is a R2R issue solely because R2R does have a lot of special handling for ARM32, especially around type layouts, as opposed to single file or trimming. Disabling R2R of crossgen2 for ARM32 Linux and running the CI would be a quick way to validate.)

am11 commented:
ReadyToRun disabled, it's still not clearing the CI:
```
    readytorun/tests/mainv2/mainv2.sh [FAIL]
      /root/helix/work/workitem/e/readytorun/tests/mainv2/mainv2.sh: line 214:  7391 Segmentation fault      (core dumped) $CORE_ROOT/crossgen2/crossgen2 --map -r:$CORE_ROOT/*.dll -r:IL_DLLS/fieldgetter.dll -r:IL_DLLS/mainv2.dll -o:test.dll ../testv2/test/test.dll
      
      Return code:      1
      Raw output file:      /root/helix/work/workitem/uploads/Reports/readytorun.tests/mainv2/mainv2.output.txt
      Raw output:
      BEGIN EXECUTION
      Crossgen failed with exitcode: 139
      Test failed. Trying to see if dump file was created in /home/helixbot/dotnetbuild/dumps since 2/9/2023 4:12:43 AM
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=/root/helix/work/correlation
      > /root/helix/work/workitem/e/readytorun/tests/mainv2/mainv2.sh
      Expected: True
      Actual:   False
      Stack Trace:
           at readytorun_tests._mainv2_mainv2_._mainv2_mainv2_sh()
           at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
           at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
````

I ran this test locally via qemu and it runs successfully (once ../testv2/test/test.dll is in the right place which I placed manually maybe because `src/tests/build.sh -dir:readytorun` isn't working even after https://github.com/dotnet/runtime/commit/232fbcea7f56d4e23fa5af7a0a490f59195dc5f4 applied?), but this failure is more than a FileNotFoundException.

trylek commented:
@am11, could you please update the PR by resolving the conflict? Other than that, I'm starting to think that the experiment you carried out based on Michal's suggestion by removing the <code>PublishReadyToRun</code> property nicely divided the problem space and showed that the problem is likely in the <code>PublishSingleFile</code> property that you left in the property group. I would suggest trying it the other way around - putting the <code>PublishReadyToRun</code> lines back and removing <code>PublishSingleFile</code>. If that makes the problem go away, we're probably golden for now.

trylek commented on src/tests/build.proj:
```diff
@@ -529,6 +529,8 @@
       <CrossgenDir Condition="'$(TargetArchitecture)' == 'arm64'">$(CrossgenDir)\x64</CrossgenDir>
       <CrossgenDir Condition="'$(TargetArchitecture)' == 'x86'">$(CrossgenDir)\x64</CrossgenDir>
+      <Crossgen2Path>$(CrossgenDir)\crossgen2\crossgen2</Crossgen2Path>
+      <Crossgen2Path Condition="'$(TargetOS)' == 'windows'">$(Crossgen2Path).exe</Crossgen2Path>
```
I think this is wrong, test framework gets crossgenned on the AzDO build machine so we should be checking <code>HostOS</code>, not <code>TargetOS</code> (not sure if that is actually causing any of the lab failures but I think it's still a bug).

am11 replied:
We don't support Windows<->Unix cross OS compilation, so it's debatable..

trylek commented:
FWIW, I believe the weird <code>ManifestMetadata</code> bug is caused by the fact that somehow indirectly the final composite crossgen2 dll is also subjected to the closure verification task - I checked binlogs from your run and from recent Eduardo's run

https://dev.azure.com/dnceng-public/public/_build/results?buildId=202759&view=results

and you can easily see that in the latter case <code>crossgen2</code> and <code>crossgen2.r2r</code> (the composite image) aren't present in the closure validation assembly list. <code>ManifestMetadata</code> is not a real assembly name, that refers to a technical pseudo-assembly that gets stored inside the composite image. I haven't yet managed to figure out what aspect of your change could be affecting this, I'll try to take a closer look tomorrow unless you beat me to it.

am11 committed:
Copy published crossgen2 in artifacts/tests

am11 committed:
Merge branch 'main' into feature/tests/published-crossgen2

trylek commented:
@am11 - I see that you have removed the "needs-author-action" label but I see no recent commits on the PR. Also, for merging this in we need to rerun testing which also means resolving conflicts in your change. As you can easily imagine, we're now close to shipping .NET 8 and all pending issues and PRs are regularly scrutinized, can you please share how you plan to proceed with this change so that I can speak to it in our next internal backlog meeting?

am11 commented:
> I'll try to take a closer look tomorrow unless you beat me to it.

@trylek did you get a chance to take a look at arm issue? I don't have arm hardware to investigate the arm failure.

trylek commented:
@am11 - I think that as a first step we should rebase this PR against latest main and rerun the testing, multiple ARM bugs have been fixed since your last round of testing. Please let me know if you plan to do the rebase or if I should do it.

am11 committed:
Merge dotnet/main into feature/tests/published-crossgen2

am11 commented on src/coreclr/tools/aot/crossgen2/crossgen2.csproj:
```diff
@@ -52,6 +52,11 @@
           Condition="'$(NativeAotSupported)' == 'true' and '$(_IsPublishing)' == 'true'" />
   <Import Project="$(RepositoryEngineeringDir)nativeSanitizers.targets" />
 
+  <!-- Some asserts in CoreLib are not compatible with metadata blocking -->
+  <ItemGroup>
+    <IlcArg Include="--nometadatablocking" />
+  </ItemGroup>
+
```
```suggestion

```

am11 committed:
Delete --nometadatablocking workaround

am11 commented:
@trylek, all yours. Mono failure is unrelated (https://github.com/dotnet/runtime/issues/90605). Arm failure is the same as before. There is a dump available at the bottom on [logs page](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-80154-merge-3b185233835a4af3b7/PayloadGroup0/1/console.554af45f.log?helixlogtype=result).

trylek commented:
@jkoritzinsky - Jeff asked me to help am11 drive this long ago. I've been trying to rebase the change against current main but I'm somewhat worried that it might conflict with the Crossgen2 build refactoring you worked on recently. AFAIK am11's intent was to use the NativeAOT-built Crossgen2 for running the CoreCLR tests; I think that in his PR he's calling into the packs partition to build the "final" Crossgen2. In general I like the intent to test the Crossgen2 we're really shipping but you might have additional requirements and / or limitations so we need to make sure this change doesn't go against your plans. Alternatively please let me know if you believe your recent changes supersede this PR which could be then closed as no longer needed.

Thanks Tomas

jkoritzinsky commented:
@trylek this work should still make sense with the crossgen2 build refactoring. It shouldn't need to make any changes to the crossgen2 project files. I'll add a comment in one spot for how I'd recommend writing out the "live/final" crossgen2 binary.

jkoritzinsky commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -36,6 +37,14 @@
 
+    <!-- Publish crossgen2 on supported platforms. -->
+
+    <MSBuild Condition="'$(Crossgen2Supported)' == 'true'"
+      Targets="Restore;Publish"
+      BuildInParallel="true"
+      Properties="PublishDir=$(CORE_ROOT)\crossgen2;DisableSourceLink=true"
+      Projects="$(InstallerProjectRoot)pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Crossgen2.sfxproj" />
```
If you use the PublishToDisk target, you'll get the same layout as the tarball/zip has. Additionally, if we ever make the managed test build act as a subset in the root build scripts (instead of being a separate script), the build will correctly dedupe against a build of the sfxproj directly.

```suggestion
    <MSBuild Condition="'$(Crossgen2Supported)' == 'true'"
      Targets="PublishToDisk"
      BuildInParallel="true"
      Properties="OutputPath=$(CORE_ROOT)\crossgen2"
      Projects="$(InstallerProjectRoot)pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Crossgen2.sfxproj" />
```

am11 replied:
Looks like we need to pass additional props for Publish->PublishToDisk, becasue tests are not resolving NativeAotSupported anymore. The preference in crossgen2 publishing used to be in this order: PublishAot->PublishReadyToRun->PublishSingleFile based on the target platform. That is apparently not working after the rebase (and all platforms are looking for apphost).

am11 committed:
Merge branch 'main' into feature/tests/published-crossgen2

am11 committed:
Resolve merge conflicts

am11 commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -36,6 +37,14 @@
+    <!-- Publish crossgen2 on supported platforms. -->
+
+    <MSBuild Condition="'$(Crossgen2Supported)' == 'true'"
+      Targets="PublishToDisk"
```
```suggestion
      Targets="Restore;PublishToDisk"
```

am11 committed:
Update src/tests/Common/Directory.Build.targets

agocke commented:
FWIW we now have a separate `crossgen2_publish` project that might make this easier. I'm going to move this to draft for now until someone has time to fix up the rebase.

am11 committed:
Merge branch 'main' into feature/tests/published-crossgen2

am11 commented:
With all the arm32 fixes for nativeaot merged, I think we can just enable publishing linux-arm crossgen2 to AOT directly as well.

> FWIW we now have a separate `crossgen2_publish` project that might make this easier. I'm going to move this to draft for now until someone has time to fix up the rebase.

@agocke, thanks that probably explain why all targets have stopped publishing crossgen2 as AOT app since last month in this branch: https://github.com/dotnet/runtime/pull/80154#discussion_r1462862138. I will take a closer look and enable linux-arm AOT publishing as well.

am11 committed:
Build crossgen2_publish and enable arm AOT

am11 committed:
Pass NativeAotSupported

am11 committed:
Skip apphost when PublishAot set

agocke commented:
@jkoritzinsky Mind taking a look?

jkoritzinsky commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -36,6 +37,14 @@
 
+    <!-- Publish crossgen2 on supported platforms. -->
+
+    <MSBuild Condition="'$(Crossgen2Supported)' == 'true'"
+      Targets="Restore;PublishToDisk"
+      BuildInParallel="true"
+      Properties="NativeAotSupported=$(NativeAotSupported);OutputPath=$(CORE_ROOT)\crossgen2"
+      Projects="$(InstallerProjectRoot)pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Crossgen2.sfxproj" />
```
Let's follow the same pattern elsewhere in the repo where we use a separate invocation with `MSBuildRestoreSessionId` set to a new Guid to ensure we get the correct restore experience.

jkoritzinsky reviewed (commented):
Just one comment. Once the merge conflicts are resolved and the PR is green, I'll do a final pass. Other than that, looking good!

am11 commented:
@jkoritzinsky, before #92826 only arm was failing, now all archs are looking for apphost and failing. Do you know which change in particular affected it?

jkoritzinsky commented:
My guess would be the changes to import targetingpacks.targets, but I can't verify without a binlog and the build expired.

am11 committed:
Merge dotnet/main into feature/tests/published-crossgen2

am11 committed:
Address CR feedback

am11 committed:
Merge branch 'main' into feature/tests/published-crossgen2

am11 commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -42,7 +42,7 @@
       Targets="Restore;PublishToDisk"
       BuildInParallel="true"
-      Properties="NativeAotSupported=$(NativeAotSupported);OutputPath=$(CORE_ROOT)\crossgen2"
+      Properties="NativeAotSupported=$(NativeAotSupported);OutputPath=$(CORE_ROOT)\crossgen2;MSBuildRestoreSessionId=$([System.Guid]::NewGuid());"
```
Assuming `MSBuildRestoreSessionId` applies only to restore, will split make any difference?

jkoritzinsky replied:
Yes, as it will force a re-evaluation of the MSBuild project for the `PublishToDisk` target as the global properties differ from the `Restore` target invocation.

jkoritzinsky commented on src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj:
```diff
@@ -29,6 +27,7 @@
     <!-- Disable crossgen on FreeBSD when cross building from Linux. -->
     <PublishReadyToRun Condition="'$(TargetOS)' == 'freebsd' and '$(CrossBuild)' == 'true'">false</PublishReadyToRun>
     <PublishReadyToRunComposite>$(PublishReadyToRun)</PublishReadyToRunComposite>
+    <UseLocalAppHostPack>true</UseLocalAppHostPack>
```
Can you move this line back up to the unconditional property group? I think this is what's causing the failures (as the local apphost pack can't be found).

am11 replied:
After reverting, next place where it looks for apphost is:
> .dotnet/sdk/9.0.100-preview.3.24204.13/Microsoft.Common.CurrentVersion.targets(5270,5): error MSB3030: Could not copy the file "/__w/1/s/artifacts/obj/coreclr/crossgen2_publish/linux.x64.Checked/apphost" because it was not found.
(from `Build coreclr Pri0 Runtime Tests Run linux x64 checked failed`)

MichalStrehovsky replied:
Is crossgen2_publish NativeAOT'd here?

We just updated the repo to Preview 3 SDK that has this fix: https://github.com/dotnet/sdk/pull/38644 - SDK should not even be looking for an apphost when doing NAOT publish.

am11 replied:
I tried a few things but this looks like a deadlock.
* crossgen2_publish project is manully publishing AOT app using live build, meaning PublishAot=true is not set
* explicitly setting UseAppHost=false with SelfContained=false for NativeAotSupported != '' failed with `error NETSDK1102: Optimizing assemblies for size is not supported for the selected publish configuration. Please ensure that you are publishing a self-contained app`
* explicitly setting SelfContained=false with with SelfContained=true failed with `error NETSDK1067: Self-contained applications are required to use the application host. Either set SelfContained to false or set UseAppHost to true.`
* also PublishTrimmed is requiring apphost

am11 replied:
Yes, this manual setup is likely there so crossgen2 is published with live build of ILCompiler when NativeAotSupported==true. ILCompiler itself is published using nuget package. crossgen2_publish.csproj line 4 is setting `_IsPublishing`. Invoking `Publish` / `PublishToDisk` directory gives the same missing apphost error. -v:diag dump suggests one of the linker/trimmer target is trying to copy apphost where it fails.

am11 committed:
Update crossgen2_publish.csproj

am11 committed:
Add `alpine` to target triplet for crossbuild

am11 committed:
Update Microsoft.NETCore.Native.Unix.targets

am11 commented:
> Add alpine to target triplet for crossbuild

We will need to wait for next preview with https://github.com/dotnet/runtime/pull/101213. Locally, I've verified that it fixed the  alpine-arm crossbuild by copying the live version to nuget:

```
cp /runtime/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets \
   /root/.nuget/packages/microsoft.dotnet.ilcompiler/9.0.0-preview.4.24215.1/build
```

am11 committed:
Use_OrNotToUse_AppHost

am11 commented on src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj:
```diff
@@ -17,6 +16,7 @@
     <SelfContained>true</SelfContained>
```
```suggestion
    <SelfContained Condition="'$(NativeAotSupported)' != 'true'>true</SelfContained>
```

am11 committed:
Update src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj

am11 committed:
Merge branch 'main' into feature/tests/published-crossgen2

am11 commented on src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj:
```diff
@@ -8,15 +8,15 @@
     <PublishTrimmed>true</PublishTrimmed>
     <RuntimeIdentifier>$(PackageRID)</RuntimeIdentifier>
-    <SelfContained>true</SelfContained>
+    <SelfContained Condition="'$(NativeAotSupported)' != 'true'>true</SelfContained>
```
```suggestion
    <SelfContained Condition="'$(NativeAotSupported)' != 'true'">true</SelfContained>
```

am11 committed:
.

am11 commented on src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj:
```diff
@@ -8,15 +8,15 @@
     <PublishTrimmed>true</PublishTrimmed>
     <RuntimeIdentifier>$(PackageRID)</RuntimeIdentifier>
-    <SelfContained>true</SelfContained>
+    <SelfContained Condition="'$(NativeAotSupported)' != 'true'">true</SelfContained>
```
```suggestion
    <SelfContained Condition="'$(NativeAotSupported)' == 'true'">false</SelfContained>
    <SelfContained Condition="'$(NativeAotSupported)' != 'true'">true</SelfContained>
```

am11 committed:
Update src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj

am11 committed:
Revert apphost experiment

am11 committed:
Split <MSBuild> task invocation for restore/publish

am11 commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -36,6 +37,20 @@
+      Projects="$(InstallerProjectRoot)pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Crossgen2.sfxproj" />
+
+    <MSBuild Condition="'$(Crossgen2Supported)' == 'true'"
+      Targets="PublishToDisk"
+      BuildInParallel="true"
+      Properties="NativeAotSupported=$(NativeAotSupported);OutputPath=$(CORE_ROOT)\crossgen2;"
```
```suggestion
      Targets="Build"
      BuildInParallel="true"
      Properties="NativeAotSupported=$(NativeAotSupported);PublishDir=$(CORE_ROOT)\crossgen2;"
```

am11 committed:
Update src/tests/Common/Directory.Build.targets

am11 commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -36,6 +37,20 @@
+      Projects="$(InstallerProjectRoot)pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Crossgen2.sfxproj" />
+
+    <MSBuild Condition="'$(Crossgen2Supported)' == 'true'"
+      Targets="Build"
```
```suggestion
      Targets="Publish"
```

am11 committed:
Update src/tests/Common/Directory.Build.targets

am11 committed:
Merge branch 'main' into feature/tests/published-crossgen2

am11 commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -36,6 +37,20 @@
+      Projects="$(InstallerProjectRoot)pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Crossgen2.sfxproj" />
+
+    <MSBuild Condition="'$(Crossgen2Supported)' == 'true'"
+      Targets="Publish"
+      BuildInParallel="true"
+      Properties="NativeAotSupported=$(NativeAotSupported);PublishDir=$(CORE_ROOT)\crossgen2;"
```
```suggestion
      Targets="PublishToDisk"
      BuildInParallel="true"
      Properties="NativeAotSupported=$(NativeAotSupported);OutputPath=$(CORE_ROOT)\crossgen2;"
```

am11 committed:
revert

ivdiazsa reviewed (commented):
At a glance, this looks good. It would be a good idea to revive this PR and test accordingly with today's codebase.

am11 committed:
Merge branch 'main' into feature/tests/published-crossgen2

ivdiazsa commented:
/azp run runtime-coreclr outerloop

am11 committed:
Simplify Crossgen2Runner

am11 committed:
Add missing overrides

am11 committed:
Account for new properties

am11 committed:
Require apphost in non-Aot branch

am11 committed:
Keep requiring apphost

am11 commented:
@jkoritzinsky, I have tried two variations of `UseLocalAppHostPack` in crossgen2_publish.csproj:

1. `UseLocalAppHostPack=true` as it is in `main`: it passes the `host` subset build but fails to find apphost when building `src/tests/build.sh generatelayoutonly` when runtime (and thereby host) configuration is different than library configuration (release,debug or checked,release etc.)
2. `UseLocalAppHostPack` is set to `true` for `Condition="'$(UseNativeAotForComponents)' != 'true'"` and `false` for `Condition="'$(UseNativeAotForComponents)' == 'true'"` in crossgen2_publish.proj: it passes the `src/tests/build.sh` but fails to find apphost for target platform when building `packs` subset, i.e. when CrossBuild is true.

This problem started to happen around merge at this point: https://github.com/dotnet/runtime/pull/80154#issuecomment-2060649529. I think a deeper dive into internal SDK targets maybe needed to fix this disparity:
<img width="1068" alt="image" src="https://github.com/user-attachments/assets/178d7591-05aa-4a94-80af-a26dfc9fc09e">

but if you know of a cleaner, more idiomatic, solution I'd rather vote for that one to avoid more custom targets.

jkoritzinsky commented:
@am11 you might need to flow the `HostConfiguration` property through if it's only hitting in different-host-config scenarios.

I think that the solution for item 1 would be to require the user to set a host config option if they want something other than Release.

Unless you only need the single-file-host, at which point you can set `HostConfiguration` to the runtime config (instead of the libraries config).

Is there a way we can make sure this dev workflow change only happens for crossgen2 testing? It would be nice to not break everyones workflow by requiring them to build the hosts when they're not going to need them.

am11 commented:
> I think that the solution for item 1 would be to require the user to set a host config option if they want something other than Release.

Good point! I will try that. Even when osx is happy, linux starts to fail because --cross in osx and linux differ. Hopefully, it will be sorted out.

> Is there a way we can make sure this dev workflow change only happens for crossgen2 testing? It would be nice to not break everyones workflow by requiring them to build the hosts when they're not going to need them.

I'm aiming to maintain the existing supported scenarios without requiring user to do anything special. CI has a pretty good coverage of most common scenarios. There are cases like https://github.com/dotnet/runtime/pull/105004, where we may want to build ilcompiler project (as .dll; stage 0), then use that to PublishAot the ILCompiler project (stage 1), then use that to PublishAot it again (stage 2) to complete the loop for all official/community platforms alike, but right now with apphost requirement for PublishAot and PublishSingleFile is making things a bit tricky to reason about.

am11 committed:
Manually copy apphost to crossgen2 obj dir

am11 committed:
Merge dotnet/main into feature/tests/published-crossgen2

am11 committed:
Fix paths

am11 committed:
Account for tools configurations

am11 committed:
Remove redundant double quotes

am11 commented on src/tests/Common/CLRTest.CrossGen.targets:
```diff
@@ -243,15 +244,14 @@ if defined RunCrossGen2 (
     set __R2RDumpCommand=!__R2RDumpCommand! --header --sc --in !__OutputFile! --out !__OutputFile!.r2rdump --val
-    set __Command=!__Command! "!CORE_ROOT!\crossgen2\crossgen2.dll"
+
+    set __Command=!_DebuggerFullPath! "!CORE_ROOT!\crossgen2\crossgen2.exe"
```
```suggestion
    set __Command=!_DebuggerFullPath! "!CORE_ROOT!\crossgen2\crossgen2.exe"
    for /r "%CORE_ROOT%" %%i in (System*.dll) do @echo %%i
```

am11 committed:
Update src/tests/Common/CLRTest.CrossGen.targets

am11 commented on src/tests/Common/CLRTest.CrossGen.targets:
```diff
@@ -243,15 +244,15 @@ if defined RunCrossGen2 (
-    set __Command=!__Command! "!CORE_ROOT!\crossgen2\crossgen2.dll"
+
+    set __Command=!_DebuggerFullPath! "!CORE_ROOT!\crossgen2\crossgen2.exe"
+    for /r "%CORE_ROOT%" %%i in (System*.dll) do @echo %%i
```
```suggestion
    py.exe -m pip install pefile
    py -c "import pefile, sys; dll_path = sys.argv[1]; pe = pefile.PE(dll_path); [print(entry.dll.de
code('utf-8')) for entry in pe.DIRECTORY_ENTRY_IMPORT]" %CORE_ROOT%\System.Runtime.dll
```

am11 committed:
Update src/tests/Common/CLRTest.CrossGen.targets

am11 committed:
Update src/tests/Common/CLRTest.CrossGen.targets

am11 commented on src/tests/Common/CLRTest.CrossGen.targets:
```diff
@@ -243,15 +244,16 @@ if defined RunCrossGen2 (
-    set __Command=!__Command! "!CORE_ROOT!\crossgen2\crossgen2.dll"
+
+    set __Command=!_DebuggerFullPath! "!CORE_ROOT!\crossgen2\crossgen2.exe"
+    py.exe -m pip install pefile
+    py.exe -c "import pefile, sys; dll_path = sys.argv[1]; pe = pefile.PE(dll_path); [print(entry.dll.decode('utf-8')) for entry in pe.DIRECTORY_ENTRY_IMPORT]" %CORE_ROOT%\System.Runtime.dll
```
```suggestion
    py.exe -m pip install pefile
    py.exe -c "import pefile, sys%3B dll_path = sys.argv[1]%3B pe = pefile.PE(dll_path)%3B [print(entry.dll.decode('utf-8')) for entry in pe.DIRECTORY_ENTRY_IMPORT]" %CORE_ROOT%\System.Runtime.dll
```

am11 committed:
Update src/tests/Common/CLRTest.CrossGen.targets

am11 commented on src/tests/Common/CLRTest.CrossGen.targets:
```diff
@@ -243,15 +244,16 @@ if defined RunCrossGen2 (
-    set __Command=!__Command! "!CORE_ROOT!\crossgen2\crossgen2.dll"
+
+    set __Command=!_DebuggerFullPath! "!CORE_ROOT!\crossgen2\crossgen2.exe"
+    py.exe -m pip install pefile
+    py.exe -c "import pefile, sys%3B dll_path = sys.argv[1]%3B pe = pefile.PE(dll_path)%3B [print(entry.dll.decode('utf-8')) for entry in pe.DIRECTORY_ENTRY_IMPORT]" %CORE_ROOT%\System.Runtime.dll
```
```suggestion
```

am11 committed:
Update src/tests/Common/CLRTest.CrossGen.targets

am11 commented:
@jkoritzinsky, manually copying the apphost fixed the issue. For:

* `./build.sh clr+libs -rc Checked`
* `./build.sh clr+libs -rc Checked -c Release`

the respective test commands should be:
* `src/tests/build.sh -p:RuntimeConfiguration=Checked`
* `src/tests/build.sh -Release -p:RuntimeConfiguration=Checked`

However, since we have the defaults flipped, we are using:

* `src/tests/build.sh -Checked -p:LibrariesConfiguration=Debug`
* `src/tests/build.sh -Checked -p:LibrariesConfiguration=Release`

What do you think about fixing this disparity? (e.g. as a separate PR)
For now, I have added a workaround for it with a comment in `src/tests/Common/Directory.Build.targets`.

---

Now unix tests are building and executing with published crossgen2. Windows tests are also building but one test is failing only in CI, `crossgen2smoke.cmd`:

```
19:53:58.146 Running test: readytorun/crossgen2/crossgen2smoke/crossgen2smoke.cmd

Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
```

I tested with pip package `pefile` in CI (which I've now reverted in a9e539b), and apparently there is no obvious "unmet" dependency issue, file is there and System.Runtime.dll depends on mscoree.dll https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-80154-merge-5000519e1f4446eda0/readytorun/1/console.344b2ad0.log?helixlogtype=result

I built it on windows 11 machine the same way as CI does (-rc checked -c debug) and that test passes. Any ideas what to do with it? Note that CLRTest.CrossGen.targets stamps out that cmd file which is used by multiple tests, but only crossgen2smoke.cmd is failing in the CI.

am11 committed:
Merge branch 'main' into feature/tests/published-crossgen2

jkoritzinsky commented:
I don't want to change how the configurations work right now as the tests build script defaults are better aimed at the people who will be using it (people working on runtimes, not libraries or host). If we ever get to the point of hooking everything to the root build script, then we'll normalize it.

I'll take a look at the failure when I have a sec.

am11 committed:
Keep defining __DotNet

am11 committed:
Merge branch 'main' into feature/tests/published-crossgen2

am11 commented:
Nvm, I've figured it out. It was a silly oversight that __DotNet was used in part of that batch script which the smoke test exercises.

am11 commented:
@jkoritzinsky, @MichalStrehovsky, @agocke, PTAL. I know it only took 20 months to get it to a working state but there were other PRs by multiple people which contribute to its development over time. 😅

ivdiazsa reviewed (approved):
This lgtm thank you very much for your contribution for this @am11! Let's wait and hear the other reviewers' comments, as they are more familiarized with this infrastructure component than I am.

jkoritzinsky commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -34,7 +35,47 @@
-          DependsOnTargets="ResolveAssemblyReferences;ResolveRuntimeFilesFromLocalBuild">
+    DependsOnTargets="ResolveAssemblyReferences;ResolveRuntimeFilesFromLocalBuild">
+
+    <!-- Copy apphost to crossgen2_publish directory.
+        The default configuration between product and tests are flipped. When we build project like:
+        `build -c debug -rc checked`, the corresponding test command is `src/tests/build -checked -p:LibrariesConfiguration=debug`,
+        instead of `-debug -p:RuntimeConfiguration=checked`. That forces us to either pass the
+        `HostConfiguration=debug` explicitly or fix this disparity; both of which will break the dev workflow.
+        
+        As a workaround, we will first check if the directory pointed by `HostConfiguration` exists, then check
+        `LibrariesConfiguration`.
+    -->
```
How will this workaround handle a local workflow like `build clr+libs -c debug -lc release`? Will it still work correctly? Or do we need to also check the apphost built with `RuntimeConfiguration`?

am11 replied:
Yup, it’s basically tested in the CI as well. What happens is that HostConfiguration is resolved to debug (because the un-prefixed Configuration is debug). But we then build the tests with opposite `src/tests/build.sh release -LibrariesConfiguration=Debug`, which requires this reassignments (because now HostConfiguration will be resolved to release instead of debug). As you mentioned above, once we integrate test build script with top-level build.sh/cmd, we can remove this workaround then.

jkoritzinsky commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -34,7 +35,47 @@
+
+    <MakeDir Condition="'$(Crossgen2Supported)' == 'true'" Directories="$(ArtifactsObjDir)coreclr\crossgen2_publish\$(TargetOS).$(TargetArchitecture).$(RuntimeConfiguration)" />
+
+    <Copy
+      SourceFiles="$(_apphostPath)"
+      DestinationFiles="$(ArtifactsObjDir)coreclr\crossgen2_publish\$(TargetOS).$(TargetArchitecture).$(RuntimeConfiguration)\apphost$(ExeSuffix)"
+      Condition="'$(Crossgen2Supported)' == 'true'" />
```
Instead of copying this into place, could we instead pass `HostConfiguration` as a property below like `ToolsConfiguration` with a value that will resolve to an apphost that has been built?

am11 replied:
I tried that, it wasn’t working. It’s basically just this file `apphost.exe`, which is resolved by some ResolveFrameworkReferences type of target that a `Copy` task later in this file requires and complains about. I haven’t intensively debugged it but I think there maybe a better solution which we can find later on.

Funny thing is crossgen2 publishing is either using the singlefilehost or AOT, so so this apphost requirement is just a bogus formality to satisfy a validation check.

jkoritzinsky reviewed (commented):
A few comments around how this impacts some workflows.

Can you make a temp change in the JIT to trigger the superpmi-diffs pipeline? I want to make sure we don't break that.

jkoritzinsky commented:
Thanks for continuing to work on this even as infra has moved from under you!

am11 committed:
dummy change in jit

am11 commented:
spmi legs are green with 2065108, reverting the dummy change.

am11 commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -31,6 +31,7 @@ unsigned Compiler::jitTotalMethodCompiled = 0;
 LONG Compiler::jitNestingLevel = 0;
 #endif // defined(DEBUG)
 
+// dummy change to trigger SPMI CI legs
```
```suggestion
```

am11 committed:
revert dummy change

am11 committed:
Merge branch 'main' into feature/tests/published-crossgen2

jkoritzinsky commented:
Let's run one more crossgen outerloop pipeline before merging, but otherwise looks great to me!

jkoritzinsky commented:
/azp run runtime-coreclr crossgen2

jkoritzinsky commented:
@am11 can you take a look at the failures in the runtime-coreclr crossgen2 pipeline? Once we get those fixed, I think this is ready to merge.

am11 committed:
Use coreroot in CrossgenFramework target

am11 commented:
Good call @jkoritzinsky, there was this remaining crossgen-path using non-published, apparently only used by `src/tests/build.sh crossgen2 ...`. Could you rerun that leg again?

jkoritzinsky commented:
/azp run runtime-coreclr crossgen2

am11 commented:
Possible follow-up work: we have

```
runtime/src/tasks/Crossgen2Tasks/PrepareForReadyToRunCompilation.cs
runtime/src/tasks/Crossgen2Tasks/ResolveReadyToRunCompilers.cs
runtime/src/tasks/Crossgen2Tasks/RunReadyToRunCompiler.cs
```

duplicate of

```
sdk/src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs
sdk/src/Tasks/Microsoft.NET.Build.Tasks/ResolveReadyToRunCompilers.cs
sdk/src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs
```

both are using crossgen2.dll for some reason (run `git grep '<Crossgen2Tool'`). We should switch them to use published `crossgen2` as well if it is possible. At the same time, deduplicate them (i.e. remove the SDK copy) and remove old `crossgen one` support from them as well, if possible?

am11 committed:
Support cross CG2 which is not running on helix

am11 commented:
Cross built tests payload is sent to helix to run on the target machine for which crossgen2.exe was built for. This works great. Then we have this one-of-a-kind `__TestBuildMode = crossgen2` thing whose payload is running on the same machine. That requires us to keep the crossgen2.dll MSIL approach: 2324602. We should send it to the same helix plan as all other R2R related tests.

am11 commented:
Cross built tests payload is sent to helix to run on the target machine for which crossgen2.exe was published. This works great. Then we have this one-of-a-kind `__TestBuildMode = crossgen2` thing whose cross-built payload is running on the same machine. That requires us to keep the crossgen2.dll MSIL approach: 2324602. We should move it to the same helix plan as all other R2R related tests.

am11 commented:
One more time `/azp run runtime-coreclr crossgen2` please. 🤞

jkoritzinsky commented:
/azp run runtime-coreclr crossgen2

jkoritzinsky commented:
It's working now. Thank you for getting this working!

am11 commented:
Thank you! BinaryFormatter failure was https://github.com/dotnet/runtime/issues/104216.

Follow up list is available here: https://github.com/dotnet/runtime/issues/106929.

jakobbotsch commented:
With this PR the repo no longer builds for me. I get this error:
```
C:\dev\dotnet\runtime2\.dotnet\sdk\9.0.100-preview.7.24407.12\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceRes
olution.targets(495,5): error NETSDK1112: The runtime pack for Microsoft.NETCore.App.Runtime.win-x64 was not downloaded. Try running
 a NuGet restore with the RuntimeIdentifier 'win-x64'. [C:\dev\dotnet\runtime2\src\coreclr\tools\aot\crossgen2\crossgen2_publish.csp
roj]
```

jakobbotsch commented:
Also, `superpmi_collect` fails during "Create Core_Root" step on linux-arm (and probably other platforms) with:
```
2024-08-25T18:07:53.9782152Z   Unhandled exception: System.DllNotFoundException: Unable to load shared library 'clrjit_universal_arm_x64' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: 
2024-08-25T18:07:53.9787522Z   /mnt/vss/_work/1/s/artifacts/bin/coreclr/linux.arm.Checked/x64/ilc/clrjit_universal_arm_x64.so: cannot open shared object file: No such file or directory
2024-08-25T18:07:53.9788209Z   /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.38' not found (required by /mnt/vss/_work/1/s/artifacts/bin/coreclr/linux.arm.Checked/x64/ilc/libclrjit_universal_arm_x64.so)
2024-08-25T18:07:53.9788573Z   libclrjit_universal_arm_x64.so: cannot open shared object file: No such file or directory
2024-08-25T18:07:53.9788916Z   /mnt/vss/_work/1/s/artifacts/bin/coreclr/linux.arm.Checked/x64/ilc/clrjit_universal_arm_x64: cannot open shared object file: No such file or directory
2024-08-25T18:07:53.9790253Z   /mnt/vss/_work/1/s/artifacts/bin/coreclr/linux.arm.Checked/x64/ilc/libclrjit_universal_arm_x64: cannot open shared object file: No such file or directory
2024-08-25T18:07:53.9790466Z   
2024-08-25T18:07:54.0025059Z      at System.Runtime.InteropServices.NativeLibrary.LoadLibraryByName(String libraryName, Assembly assembly, Nullable`1 searchPath, Boolean throwOnError)
2024-08-25T18:07:54.0025730Z      at Internal.JitInterface.JitConfigProvider.<>c__DisplayClass5_0.<Initialize>b__0(String libName, Assembly assembly, Nullable`1 searchPath) in /_/src/coreclr/tools/Common/JitInterface/JitConfigProvider.cs:line 48
2024-08-25T18:07:54.0026210Z      at System.Runtime.InteropServices.NativeLibrary.LoadLibraryCallbackStub(String libraryName, Assembly assembly, Boolean hasDllImportSearchPathFlags, UInt32 dllImportSearchPathFlags)
2024-08-25T18:07:54.0026541Z   
2024-08-25T18:07:54.0026748Z      at Internal.JitInterface.CorInfoImpl.jitStartup(IntPtr host)
2024-08-25T18:07:54.0027006Z      at Internal.JitInterface.CorInfoImpl.jitStartup(IntPtr host)
2024-08-25T18:07:54.0027324Z      at Internal.JitInterface.CorInfoImpl.Startup(CORINFO_OS os) in /_/src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:line 177
2024-08-25T18:07:54.0027706Z      at ILCompiler.RyuJitCompilationBuilder.ToCompilation() in /_/src/coreclr/tools/aot/ILCompiler.RyuJit/Compiler/RyuJitCompilationBuilder.cs:line 127
2024-08-25T18:07:54.0028041Z      at ILCompiler.Program.Run() in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 580
2024-08-25T18:07:54.0028416Z      at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass240_0.<.ctor>b__0(ParseResult result) in /_/src/coreclr/tools/aot/ILCompiler/ILCompilerRootCommand.cs:line 293
2024-08-25T18:07:54.0028690Z      at System.CommandLine.Invocation.InvocationPipeline.Invoke(ParseResult parseResult)
2024-08-25T18:07:54.0316349Z /mnt/vss/_work/1/s/artifacts/bin/coreclr/linux.arm.Checked/build/Microsoft.NETCore.Native.targets(317,5): error MSB3073: The command ""/mnt/vss/_work/1/s/artifacts/bin/coreclr/linux.arm.Checked/x64/ilc/ilc" @"/mnt/vss/_work/1/s/artifacts/obj/coreclr/crossgen2_publish/linux.arm.Checked/native/crossgen2.ilc.rsp"" exited with code 1. [/mnt/vss/_work/1/s/src/coreclr/tools/aot/crossgen2/crossgen2_publish.csproj]
2024-08-25T18:07:54.5680688Z   Determining projects to restore...
2024-08-25T18:07:55.8919801Z   All projects are up-to-date for restore.
2024-08-25T18:07:55.9387359Z 
2024-08-25T18:07:55.9388368Z Build FAILED.
```

jakobbotsch commented:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=787949&view=results also appears to have some failures that look related in the "Test crossgen2-comparison X to Y" jobs.

am11 commented:
> https://dev.azure.com/dnceng-public/public/_build/results?buildId=787949&view=results also appears to have some failures that look related in the "Test crossgen2-comparison X to Y" jobs.

Can you pinpoint the error? Those failures look to be same as https://github.com/dotnet/runtime/issues/106948.

am11 commented:
> .\build.cmd clr+libs -rc checked -c release && .\src\tests\build.cmd checked generatelayoutonly

The supported comment is (and always have been):
`.\build.cmd clr+libs -rc checked -c release && .\src\tests\build.cmd checked generatelayoutonly -p:LibrariesConfiguration=Release`

jakobbotsch commented:
> Can you pinpoint the error? Those failures look to be same as #106948.

For example: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-600889e483e24676bf/WorkItem/1/console.51e0e9bf.log?helixlogtype=result

> > .\build.cmd clr+libs -rc checked -c release && .\src\tests\build.cmd checked generatelayoutonly
> 
> The supported command is (and always have been): `.\build.cmd clr+libs -rc checked -c release && .\src\tests\build.cmd checked generatelayoutonly -p:LibrariesConfiguration=Release`

I have built this way for 3 years without issues. I do not see a mention that `LibrariesConfiguration` needs to be specified always in the workflow docs (actually I see the opposite). https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/coreclr/testing.md#building-the-core_root

am11 commented:
From [`main`/docs/workflow/testing/coreclr/testing.md#building-the-core_root](https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/coreclr/testing.md?rgh-link-date=2024-08-26T10%3A48%3A46Z#building-the-core_root)


> Note that for the libraries configuration, we are passing the argument directly to MSBuild instead of the build script, hence the `/p:LibrariesConfiguration` flag. Also, make sure you use the correct syntax depending on our platform. The cmd script takes the arguments by placing, while the sh script requires them to be with a hyphen.

am11 commented:
The fix is simple (set LibrariesConfiguration = Release by default). We did ran couple of pipelines, including superpmi by making a dummy change. Apparently it wasn't enough.

> /azp run runtime-coreclr crossgen2

We were using that as an ultimate test and didn't ran crossgen2 outerloop.

jakobbotsch commented:
> The fix is simple (set LibrariesConfiguration = Release by default).

:+1:

> We did ran couple of pipelines, including superpmi by making a dummy change. Apparently it wasn't enough.

In this case superpmi-collect is the pipeline that substantially uses crossgen2, while superpmi-diffs/superpmi-replay do not use it at all. Those latter pipelines need only JIT + SPMI builds and then download the collections created by superpmi-collect.
Sadly superpmi-collect is not public, so we (the JIT team) will have to help a bit in getting this working.

One thing that confuses me a bit is that it seems we are running ilc even when `generatelayoutonly` is specified. I would not expect `generatelayoutonly` to start running ILC and prejitting things.

> > /azp run runtime-coreclr crossgen2
> 
> We were using that as an ultimate test and didn't ran crossgen2 outerloop.

To be fair I am not sure that the crossgen2 outerloop pipeline is checked much. I just noticed it while trying to find a publicly accessible pipeline run that hit the same issue as superpmi-collect (since it is only internally accessible).

jakobbotsch commented:
> Sadly superpmi-collect is not public

I think we can add a public test-only version of this pipeline that does everything except upload the collections. That should make the testing much more straightforward for contributors.

am11 commented:
If it is blocking, lets revert it. I think two of the fixes are striaghtforward (add a line in src/tests/Common/Directory.Build.props and update crosgen2 pipeline yamls to drop .dll).

Yup, the superpmi-collect issue needs some investigation. It maybe as simple as passing --cross to:

https://github.com/dotnet/runtime/blob/5fd3f22b98cb0d126e4f6328f5a0bf922a532f69/eng/pipelines/coreclr/superpmi-collect.yml#L124-L129

or something more involved. I'll take a closer look when I am off my day job. 😅

jakobbotsch commented:
> If it is blocking, lets revert it. I think two of the fixes are striaghtforward (add a line in src/tests/Common/Directory.Build.props and update crosgen2 pipeline yamls to drop .dll).
> 
> Yup, the superpmi-collect issue needs some investigation. It maybe as simple as passing --cross to:
> 
> https://github.com/dotnet/runtime/blob/5fd3f22b98cb0d126e4f6328f5a0bf922a532f69/eng/pipelines/coreclr/superpmi-collect.yml#L124-L129
> 
> or something more involved. I'll take a closer look when I am off my day job. 😅

Thanks! I think we should revert it for now as otherwise we'll quickly end up with no SPMI collections for linux-arm/linux-arm64 while we investigate.
